app_legacy/Helpers/database/player-game.php

```diff
@@ -91,60 +91,193 @@ function getGameRankAndScore(int $gameID, string $username): array
+    if ($numRecentAchievements >= 0) {
+        usort($unlockedAchievements, function ($a, $b) {
+            if ($a['When'] == $b['When'])
+                return $a['Achievement']['ID'] <=> $b['Achievement']['ID'];
+
+            return -($a['When'] <=> $b['When']);
+        });
+
+        if ($numRecentAchievements !== 0) {
+            $unlockedAchievements = array_slice($unlockedAchievements, 0, $numRecentAchievements);
+        }
+
+        $recentAchievements = [];
+
+        foreach ($unlockedAchievements as $unlockedAchievement) {
+            $gameData = $unlockedAchievement['Game'];
+            $gameID = (int) $gameData['ID'];
+            $achievementData = $unlockedAchievement['Achievement'];
+            $achievementID = (int) $achievementData['ID'];
+
+            $recentAchievements[$gameID][$achievementID] = [
+                'ID' => $achievementID,
+                'GameID' => $gameID,
+                'GameTitle' => $gameData['Title'],
+                'Title' => $achievementData['Title'],
+                'Description' => $achievementData['Description'],
+                'Points' => (int) $achievementData['Points'],
+                'BadgeName' => $achievementData['BadgeName'],
+                'IsAwarded' => '1',
+                'DateAwarded' => $unlockedAchievement['When'],
+                'HardcoreAchieved' => (int) $unlockedAchievement['Hardcore'],
+            ];
         }
+
+        foreach ($lockedAchievements as $lockedAchievement) {
+            $gameData = $lockedAchievement['Game'];
+            $gameID = (int) $gameData['ID'];
+            $achievementData = $lockedAchievement['Achievement'];
+            $achievementID = (int) $achievementData['ID'];
+
+            $recentAchievements[$gameID][$achievementID] = [
+                'ID' => $achievementID,
+                'GameID' => $gameID,
+                'GameTitle' => $gameData['Title'],
+                'Title' => $achievementData['Title'],
+                'Description' => $achievementData['Description'],
+                'Points' => (int) $achievementData['Points'],
+                'BadgeName' => $achievementData['BadgeName'],
+                'IsAwarded' => '0',
+                'DateAwarded' => null,
+                'HardcoreAchieved' => null,
+            ];
+        }
+
+        $libraryOut['RecentAchievements'] = $recentAchievements;
     }
 
-    return $dataOut;
+    return $libraryOut;
+}
+
+function expireUserAchievementUnlocksForGame(string $user, int $gameID): void
+{
+    Cache::forget("user:$user:gameUnlocks:$gameID:" . AchievementType::OfficialCore);
+    Cache::forget("user:$user:gameUnlocks:$gameID:" . AchievementType::Unofficial);
+}
+
+function getUserAchievementUnlocksForGame(string $user, int $gameID, int $flags = AchievementType::OfficialCore): array
+{
+    return Cache::remember("user:$user:gameUnlocks:$gameID:$flags",
+        Carbon::now()->addDays(7),
```

Comment:
cache user unlocks per game for a week. primarily beneficial for recent games the user has played other than the current game the user is playing.